### PR TITLE
#206: Create directory for temporary credentials

### LINKED ISSUE #206: Startup fails when target directory does not exist
Startup fails when `target` directory does not exist.
This happens for example for non-java projects.
Proposal: 
1. exasol-testcontainers should create the folder if missing. 
2. user (other project) must / can configure the path to the file

```
org.testcontainers.containers.ContainerLaunchException: Container startup failed
        at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:349)
        at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:322)
        at com.exasol.containers.ExasolContainer.start(ExasolContainer.java:509)
        at com.exasol.exasoltestsetup.testcontainers.ExasolTestcontainerTestSetup.<init>(ExasolTestcontainerTestSetup.java:34)
        at com.exasol.exasoltestsetup.ExasolTestSetupFactory.getTestSetup(ExasolTestSetupFactory.java:48)
        at com.exasol.testsetupabstraction.server.Main.main(Main.java:37)
Caused by: com.exasol.containers.ExasolContainerInitializationException: Could not create SSH key pair
        at com.exasol.containers.ExasolContainer.getSshKeys(ExasolContainer.java:999)
        at com.exasol.containers.ExasolContainer.createDockerAccess(ExasolContainer.java:985)
        at com.exasol.containers.ExasolContainer.configure(ExasolContainer.java:186)
        at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:327)
        ... 5 more
Caused by: java.nio.file.NoSuchFileException: target/id_rsa
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
        at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)
        at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:478)
        at java.base/java.nio.file.Files.newOutputStream(Files.java:220)
        at com.exasol.containers.ssh.SshKeys$Builder.writeToFiles(SshKeys.java:121)
        at com.exasol.containers.ssh.SshKeys$Builder.createKeyPair(SshKeys.java:111)
        at com.exasol.containers.ssh.SshKeys$Builder.build(SshKeys.java:104)
        at com.exasol.containers.ExasolContainer.getSshKeys(ExasolContainer.java:997)
        ... 8 more
```

### Workaround
Create the `target` directory before starting exasol-testcontainers.

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #206.

### CODE REVIEW

File: src/main/java/com/exasol/containers/ExasolContainer.java [hunk 1]
```diff
@@ -992,14 +1004,23 @@ DockerAccess createDockerAccess() {
...
             throw new ExasolContainerInitializationException("Could not create SSH key pair", exception);
         }
     }
 
+    /**
+     * Get the path to the directory where the Exasol container stores temporary credentials.
+     *
+     * @return directory for temporary credentials
+     */
+    public Path getTemporaryCredentialsDirectory() {
```
Comment (kaklakariada) [inline src/main/java/com/exasol/containers/ExasolContainer.java:1020]:
```suggestion
    private Path getTemporaryCredentialsDirectory() {
```
I think it's not necessary to make this public (yet). Maybe in the future if it's required.

File: src/main/java/com/exasol/containers/ssh/IdentityProviderMock.java [hunk 2]
```diff
@@ -6,7 +6,7 @@
 /**
  * Enables to add an identity to a java secure channel {@link JSch}.
  */
-public interface IdentityProvider {
+public interface IdentityProviderMock {
```
Comment (kaklakariada) [inline src/main/java/com/exasol/containers/ssh/IdentityProviderMock.java:9]: Why rename this to `...Mock`? This is production code in `src/main/java`.

File: src/test/java/com/exasol/containers/ssh/DockerAccessTest.java [hunk 3]
```diff
@@ -1,47 +1,91 @@
...
+                .sessionBuilderProvider(sessionBuilderProviderMock) //
+                .dockerProbe(dockerProbeMock) //
                 .build();
-        assertThat(testee.getSshKeys(), sameInstance(sshKeys));
-        assertThat(testee.getSsh(), isA(Ssh.class));
-        assertThat(testee.getMode(), equalTo(expectedMode));
-        assertThat(testee.supportsDockerExec(), equalTo(fileExists));
+        // We intentionally ignore the result of the next call, because we are only interested in the side effect that
+        // creates the credential directory.
+        dockerAccess.getSsh();
```
Comment (kaklakariada) [inline src/test/java/com/exasol/containers/ssh/DockerAccessTest.java:87]: 👍

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: src/main/java/com/exasol/containers/ssh/IdentityProviderMock.java (hunk 2 above)
Comment (redcatbear) [reply src/main/java/com/exasol/containers/ssh/IdentityProviderMock.java:9]: Auto-refactoring gone wrong. IntelliJ was a little bit to eager here it seems.

**Review by kaklakariada — APPROVED** [review]

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**